hibernate-core/src/main/java/org/hibernate/boot/model/FunctionContributor.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.boot.model;
+
+import org.hibernate.query.sqm.function.SqmFunctionRegistry;
+import org.hibernate.service.ServiceRegistry;
+
+/**
+ * Contract for contributing functions
+ *
+ * @author Karel Maesen
+ */
+public interface FunctionContributor {
+
+	/**
+	 *  Contribute functions
+	 *
+	 * @param functionRegistry The target for the contributions
+	 * @param serviceRegistry The service registry
+	 */
+	void contributeFunctions(SqmFunctionRegistry functionRegistry, ServiceRegistry serviceRegistry);
+
+	/**
+	 * Determines order in which the contributions will be applied (lowest ordinal first).
+	 *
+	 * The range 0-500 is reserved for Hibernate, range 500-1000 for libraries and 1000-Integer.MAX_VALUE for
+	 * user-defined FunctionContributors.
+	 *
+	 * @return the ordinal for this FunctionContributor
+	 */
+	default int ordinal(){
```

Comment:
Can we just make this an argument to `#contributeFunctions`?

Reply:
This is a user/library provided service provider contract through the Java ServiceLoader mechanism, so unfortunately we can't.

Reply:
No idea what you mean here.  Why would defining the contract this way be different?

```
interface FunctionContributor {
    void contributeFunctions(int relativePosition, SqmFunctionRegistry functionRegistry, ServiceRegistry serviceRegistry);
}
```

Reply:
But again, per my other comment I personally would just drop the positioning.  You say it is deterministic, but it is not necessarily.  What happens when 2 different ones provide the same position?  Its no longer deterministic

Reply:
When there are multiple `FunctionContributor` implementations available, the order in which these contribute functions can be controlled through the `ordinal`. To make the ordering fully deterministic, the sorting uses the fully qualified name of the contributor class, but the ordinal helps with the sane ordering issues.

Reply:
Its not deterministic though.

Its the same fundamental issue as registering anything.  We used to majorly have this problem with JTA syncs e.g.

Reply:
```
class MyContributor1 implements FunctionContributor {
    ...
    @Override
    public int ordinal() { return 50; }
}

class MyContributor2 implements FunctionContributor {
    ...
    @Override
    public int ordinal() { return 50; }
}
```

Not deterministic ^^

Reply:
Yeah, that's why the sorting also includes the fully qualified class name to have a deterministic behavior. See https://github.com/hibernate/hibernate-orm/pull/4109/files#diff-568ca10f5de61f383d66055b763a8109c8d886348f70149b484aceee2294f3b5R297

But the ordinal is still useful to be able to get the ordering right.